packages/alfa-predicate/src/predicate.ts

```diff
@@ -50,6 +50,16 @@ export namespace Predicate {
     };
   }
 
+  export function and<T, U extends T, V extends U>(
+    left: Predicate<T, U>,
+    ...right: Array<Predicate<U, V>>
+  ): Predicate<T, V> {
+    return right.reduce<Predicate<T, V>>(
+      (acc, cur) => binanyAnd(acc, cur),
+      left
+    );
+  }
```

Comment:
This does actually allow callers to narrow the type `T` to a `U` and then a `V`, but will only do the last bit for the first of the `right` predicates which doesn't seem very intuitive. Also, the right predicates are currently optional so one can do `and(left)` without supplying any `right`s. That's not really an API design that we want. Here's what I'm thinking we'd want though:

```ts
function and<T, U extends T, V extends U>(
  left: Predicate<T, U>,
  right: Predicate<U, V>
): Predicate<T, V>;

function and<T>(
  left: Predicate<T>,
  right: Predicate<T>,
  ...rest: Array<Predicate<T>>
): Predicate<T>;

function and<T>(
  ...predicates: Array<Predicate<T>>
): Predicate<T> {
  return (value) => predicates.reduce<boolean>(
    (holds, predicate) => holds && predicate(value),
    true
  );
}
```

This does two things:

1. Ensure that at least two predicates are passed.
2. Only perform narrowing when a full chain can be established, such as `T -> U -> V`. When it can't, no narrowing is performed.

We could even extend this with overloads for 2+ predicates, allowing for even longer narrowing chains. If variadic kinds are ever implemented in TypeScript, we could then rely on these to get rid of the overloads.

Reply:
...and let's do the same for `or()` now that we're at it 😄

Reply:
Yes, I did consider having `and(left, right, ...rest)` and somehow discarded it 🤷